tests/components/modbus/test_init.py

```diff
@@ -177,3 +185,149 @@ async def test_config_multiple_modbus(hass, caplog):
+        assert mock_pb.write_registers.call_args[0] == (
+            data[ATTR_ADDRESS],
+            data[ATTR_VALUE],
+        )
+        mock_pb.write_registers.side_effect = ModbusException("fail write_")
+        await hass.services.async_call(
+            DOMAIN, SERVICE_WRITE_REGISTER, data, blocking=True
+        )
+
+
+async def test_pb_service_write_coil(hass, caplog):
+    """Run test for service write_coil."""
+
+    conf_name = "myModbus"
+    config = {
+        DOMAIN: [
+            {
+                CONF_TYPE: "tcp",
+                CONF_HOST: "modbusTestHost",
+                CONF_PORT: 5501,
+                CONF_NAME: conf_name,
+            }
+        ]
+    }
+
+    mock_pb = mock.MagicMock()
+    with mock.patch(
+        "homeassistant.components.modbus.modbus.ModbusTcpClient", return_value=mock_pb
+    ):
+        assert await async_setup_component(hass, DOMAIN, config) is True
+        await hass.async_block_till_done()
+
+        data = {ATTR_HUB: conf_name, ATTR_UNIT: 17, ATTR_ADDRESS: 16, ATTR_STATE: False}
+        await hass.services.async_call(DOMAIN, SERVICE_WRITE_COIL, data, blocking=True)
+        assert mock_pb.write_coil.called
+        assert mock_pb.write_coil.call_args[0] == (
+            data[ATTR_ADDRESS],
+            data[ATTR_STATE],
+        )
+        mock_pb.write_coil.side_effect = ModbusException("fail write_")
+        await hass.services.async_call(DOMAIN, SERVICE_WRITE_COIL, data, blocking=True)
+
+        data[ATTR_STATE] = [True, False, True]
+        await hass.services.async_call(DOMAIN, SERVICE_WRITE_COIL, data, blocking=True)
+        assert mock_pb.write_coils.called
+        assert mock_pb.write_coils.call_args[0] == (
+            data[ATTR_ADDRESS],
+            data[ATTR_STATE],
+        )
+
+        caplog.set_level(logging.DEBUG)
+        caplog.clear
+        mock_pb.write_coils.side_effect = ModbusException("fail write_")
+        await hass.services.async_call(DOMAIN, SERVICE_WRITE_COIL, data, blocking=True)
+        assert caplog.records[-1].levelname == "ERROR"
+        await hass.services.async_call(DOMAIN, SERVICE_WRITE_COIL, data, blocking=True)
+        assert caplog.records[-1].levelname == "DEBUG"
+
+
+async def test_pymodbus_constructor_fail(hass, caplog):
+    """Run test for failing pymodbus constructor."""
+    config = {
+        DOMAIN: [
+            {
+                CONF_TYPE: "tcp",
+                CONF_HOST: "modbusTestHost",
+                CONF_PORT: 5501,
+            }
+        ]
+    }
+    with mock.patch(
+        "homeassistant.components.modbus.modbus.ModbusTcpClient"
+    ) as mock_pb:
+        caplog.set_level(logging.ERROR)
+        mock_pb.side_effect = ModbusException("test no class")
+        assert await async_setup_component(hass, DOMAIN, config) is True
+        await hass.async_block_till_done()
+        assert len(caplog.records) == 1
+        assert caplog.records[0].levelname == "ERROR"
```

Comment:
Maybe also assert that there was a call made to `mock_pb`?

Reply:
Very good idea, done (and rebased to newest dev).